### ISSUE TITLE
Vectorize leaks (migrated from Trac #570)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#1177]
Original ticket http://projects.scipy.org/numpy/ticket/570
Reported 2007-08-13 by atmention:alberts, assigned to unknown.

http://projects.scipy.org/pipermail/numpy-discussion/2007-August/028860.html

**Comment by thouis** [comment]
Comment in Trac by atmention:teoliphant, 2007-09-21

At first blush, this looks like an expensive leak to track down.  I don't think there will be time to get to it before 1.0.4

**Comment by thouis** [comment]
Comment in Trac by atmention:stefanv, 2007-09-21

Chris, I failed to reproduce the behaviour you describe.  Would you (or anyone else following) mind providing a code snippet that exposes the bug?

I started with:

```
import numpy as N
from math import exp

invlogit = lambda x: 1.0 / (1.0 + exp(-1.0 * x))
vi = N.vectorize(invlogit)

for i in xrange(100000):
    x = vi(N.array([1,2,3]))
```

**Comment by thouis** [comment]
Comment in Trac by atmention:stefanv, 2007-10-07

More vectorize fun:

```
from numpy import *
vectorize(lambda x: "%5.3g" % x)(ones((2,2,2)))

==15041== Invalid write of size 1
==15041==    at 0x4024429: memcpy (mc_replace_strmem.c:406)
==15041==    by 0x4948A20: STRING_setitem (arraytypes.inc.src:400)
==15041==    by 0x4949188: OBJECT_to_STRING (arraytypes.inc.src:792)
==15041==    by 0x495A33C: PyArray_CastTo (arrayobject.c:7734)
==15041==    by 0x495A49E: PyArray_CastToType (arrayobject.c:7507)
==15041==    by 0x4963DA8: array_cast (arraymethods.c:667)
==15041==    by 0x80C8EEB: PyEval_EvalFrameEx (in /usr/bin/python2.5)
==15041==    by 0x80CA114: PyEval_EvalCodeEx (in /usr/bin/python2.5)
==15041==    by 0x81123C0: (within /usr/bin/python2.5)
==15041==    by 0x805C9E6: PyObject_Call (in /usr/bin/python2.5)
```

**Comment by thouis** [comment]
Comment in Trac by atmention:teoliphant, 2007-10-09

The following code leaks memory. 

```
def toto(x):
     return x**2

tutu=vectorize(toto)

Nbins=10000
for i in xrange(1000):
     c=tutu(arange(Nbins))
```

**Comment by thouis** [comment]
Comment in Trac by trac user jaberg, 2008-01-14

Travis advised me to add the following code snippet to this ticket because it is related.  It doesn't use vectorize, but it uses frompyfunc.

```
import numpy as N

l = N.random.rand(10, 10)
c = l * 1.0
f = N.frompyfunc(lambda x: 2*x, 1, 1)
while True:
    c[:] = f(l)
```

**Comment by thouis** [comment]
Comment in Trac by trac user matthieu.brucher, 2008-03-20

This thread on Python list : http://mail.python.org/pipermail/python-list/2003-June/208452.html might be related (this is called in PyUFunc_On_Om() that is the main vectorize function)

**Comment by thouis** [comment]
Comment in Trac by atmention:teoliphant, 2008-03-22

It looks like the fix in r4912 (which fixed ticket #702) also fixed this leak.  Which makes sense because vectorize makes heavy use of casting back and forth from object arrays.
